src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp

```diff
@@ -23332,6 +23334,339 @@ struct DynamicPadToPad
   }
 };
 
+struct RemoveNoOpsFromWhileLoop
+    : public CheckedOpRewritePattern<stablehlo::WhileOp,
+                                     RemoveNoOpsFromWhileLoop> {
+  using CheckedOpRewritePattern<
+      stablehlo::WhileOp, RemoveNoOpsFromWhileLoop>::CheckedOpRewritePattern;
+
+  LogicalResult matchAndRewriteImpl(stablehlo::WhileOp whileOp,
+                                    PatternRewriter &rewriter) const {
+    auto info = WhileLoopInfo(whileOp);
+    auto computeInfoSuccess = info.computeInfo();
+    if (computeInfoSuccess.failed())
+      return computeInfoSuccess;
+
+    if (!info.isValid() || !info.isConstant())
+      return failure();
+
+    auto &whileBody = whileOp.getBody().front();
+    auto inductionVar = whileBody.getArgument(0);
+
+    bool anyNoOpRemoved = false;
+
+    auto limit = info.getConstantLimit().value();
+    auto start = info.getConstantStart().value();
+    auto step = info.getConstantStep().value();
+
+    auto inductionType = inductionVar.getType();
+    unsigned bitWidth = 64;
+    if (auto tensorType = dyn_cast<RankedTensorType>(inductionType)) {
+      if (auto intType = dyn_cast<IntegerType>(tensorType.getElementType())) {
+        bitWidth = intType.getWidth();
+      }
+    }
+
+    // Initialize bounds map with induction variable bounds
+    DenseMap<Value, Bounds> boundsMap;
```

Comment:
min/max only applies if step >= 0, if negative this needs to be changed

Reply:
the induction var is initialized as 

```cpp
    APInt minBound(bitWidth, std::min(start, limit), true);
    APInt maxBound(bitWidth, std::max(start, limit), true);
    boundsMap[inductionVar] = Bounds(minBound, maxBound);
```

Reply:
hm, is min/max inclusive or exclusive then. If step > 0, range is [start, limit). If negative it's (limit, start]